File: core/src/main/java/com/toocol/ssh/core/cache/SshSessionCache.java
package com.toocol.ssh.core.cache;

import com.jcraft.jsch.ChannelShell;
import com.jcraft.jsch.Session;
import com.toocol.ssh.core.ssh.core.SshSession;
import com.toocol.ssh.utilities.functional.Switchable;

import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/3/31 12:12
 */
public class SshSessionCache {

    private static final SshSessionCache INSTANCE = new SshSessionCache();
    /**
     * the map stored all alive ssh session(include Session and ChannelShell)
     */
    private final Map<Long, SshSession> sshSessionMap = new ConcurrentHashMap<>();

    private SshSessionCache() {
    }

    public static SshSessionCache getInstance() {
        return INSTANCE;
    }

    public static int getAlive() {
        return INSTANCE.sshSessionMap.entrySet().stream()
                .filter(entry -> entry.getValue().alive())
                .toList()
                .size();
    }

    public synchronized Map<Long, SshSession> getSessionMap() {
        return sshSessionMap;
    }

    public boolean isAlive(String ip) {
        long sessionId = containSession(ip);
        if (sessionId == 0) {
            return false;
        }
        return sshSessionMap.get(sessionId).alive();
    }

    public long containSession(String ip) {
        return sshSessionMap.entrySet().stream()
                .filter(entry -> ip.equals(entry.getValue().getHost()))
                .map(Map.Entry::getKey)
                .findAny()
                .orElse(0L);
    }

    public boolean containSessionId(long sessionId) {
        return sshSessionMap.containsKey(sessionId);
    }

    public boolean isDisconnect(long sessionId) {
        if (!sshSessionMap.containsKey(sessionId)) {
            return false;
        }
        return !sshSessionMap.get(sessionId).alive();
    }

    public void putSshSession(Long sessionId, SshSession session) {
        sshSessionMap.put(sessionId, session);
    }

    public void setSession(Long sessionId, Session session) {
        sshSessionMap.computeIfPresent(sessionId, (k, v) -> {
            v.setSession(session);
            return v;
        });
    }

    public void setChannelShell(Long sessionId, ChannelShell channelShell) {
        sshSessionMap.computeIfPresent(sessionId, (k, v) -> {
            v.setChannelShell(channelShell);
            return v;
        });
    }

    public Session getSession(Long sessionId) {
        return Optional.ofNullable(sshSessionMap.get(sessionId)).map(SshSession::getSession).orElse(null);
    }

    public ChannelShell getChannelShell(Long sessionId) {
        return Optional.ofNullable(sshSessionMap.get(sessionId)).map(SshSession::getChannelShell).orElse(null);
    }

    public synchronized void stopChannelShell(long sessionId) {
        sshSessionMap.computeIfPresent(sessionId, (k, v) -> {
            v.stopChannelShell();
            return v;
        });
    }

    public synchronized void stop(long sessionId) {
        sshSessionMap.computeIfPresent(sessionId, (k, v) -> {
            v.stop();
            return null;
        });
    }

    public synchronized void stop(String host) {
        long sessionId = containSession(host);
        sshSessionMap.computeIfPresent(sessionId, (k, v) -> {
            v.stop();
            return null;
        });
        ShellCache.getInstance().stop(sessionId);
    }

    public synchronized void stopAll() {
        sshSessionMap.forEach((k, v) -> v.stop());
    }

    public Collection<Switchable> getAllSwitchable() {
        return sshSessionMap.values()
                .stream()
                .map(sshSession -> (Switchable) sshSession)
                .toList();
    }
}

File: core/src/main/java/com/toocol/termio/core/cache/SshSessionCache.kt
package com.toocol.termio.core.cache

import com.jcraft.jsch.ChannelShell
import com.jcraft.jsch.Session
import com.toocol.termio.core.ssh.core.SshSession
import com.toocol.termio.utilities.functional.Switchable
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/3/31 12:12
 */
class SshSessionCache private constructor() {
    companion object Instance {
        /**
         * the map stored all alive ssh session(include Session and ChannelShell)
         */
        private val sshSessionMap: AbstractMap<Long, SshSession> = ConcurrentHashMap()

        @get:Synchronized
        val sessionMap: Map<Long, SshSession>
            get() = sshSessionMap

        fun getAlive(): Int {
            return sshSessionMap.entries.stream()
                .filter { (_, value): Map.Entry<Long?, SshSession> -> value.alive() }
                .toList()
                .size
        }

        fun isAlive(ip: String): Boolean {
            val sessionId = containSession(ip)
            return if (sessionId == 0L) {
                false
            } else sshSessionMap[sessionId]!!.alive()
        }

        fun containSessionId(sessionId: Long): Boolean {
            return sshSessionMap.containsKey(sessionId)
        }

        fun containSession(ip: String): Long {
            return sshSessionMap.entries.stream()
                .filter { (_, value): Map.Entry<Long, SshSession> -> ip == value.host }
                .map { (key, _) -> key }
                .findAny()
                .orElse(0L)
        }

        fun isDisconnect(sessionId: Long): Boolean {
            return if (!sshSessionMap.containsKey(sessionId)) {
                false
            } else !sshSessionMap[sessionId]!!.alive()
        }

        fun putSshSession(sessionId: Long, session: SshSession) {
            sshSessionMap[sessionId] = session
        }

        fun setSession(sessionId: Long, session: Session?) {
            sshSessionMap.computeIfPresent(sessionId) { _: Long?, v: SshSession ->
                v.session = session
                v
            }
        }

        fun setChannelShell(sessionId: Long, channelShell: ChannelShell?) {
            sshSessionMap.computeIfPresent(sessionId) { _: Long?, v: SshSession ->
                v.channelShell = channelShell
                v
            }
        }

        fun getSession(sessionId: Long): Session? {
            return Optional.ofNullable(sshSessionMap[sessionId]).map { obj: SshSession -> obj.session }
                .orElse(null)
        }

        fun getChannelShell(sessionId: Long): ChannelShell? {
            return Optional.ofNullable(sshSessionMap[sessionId]).map { obj: SshSession -> obj.channelShell }
                .orElse(null)
        }

        @Synchronized
        fun stopChannelShell(sessionId: Long) {
            sshSessionMap.computeIfPresent(sessionId) { _: Long?, v: SshSession ->
                v.stopChannelShell()
                v
            }
        }

        @Synchronized
        fun stop(sessionId: Long) {
            sshSessionMap.computeIfPresent(sessionId) { _: Long?, v: SshSession ->
                v.stop()
                null
            }
        }

        @Synchronized
        fun stop(host: String) {
            val sessionId = containSession(host)
            sshSessionMap.computeIfPresent(sessionId) { _: Long?, v: SshSession ->
                v.stop()
                null
            }
            ShellCache.stop(sessionId)
        }

        @Synchronized
        fun stopAll() {
            sshSessionMap.forEach { (_: Long?, v: SshSession) -> v.stop() }
        }

        val allSwitchable: Collection<Switchable?>
            get() = sshSessionMap.values
                .stream()
                .map { sshSession: SshSession? -> sshSession }
                .toList()
    }
}